File: Sources/CountriesFlagsObjC/source/Countries/UIImage+MarshallIslands.m
//
//  UIImage+MarshallIslands.m
//  CountriesFlags
//

#import "Countries/UIImage+MarshallIslands.h"

#import "Extensions/UIBezierPath+Star.h"

@implementation UIImage (MarshallIslands)

+ (UIImage * _Nullable)flagMarshallIslandsWithSize:(CGSize)size {
    UIColor* colorBlue = [UIColor colorWithRed:0.0f
                                         green:56.0f/255.0f
                                          blue:147.0f/255.0f
                                         alpha:1.0f];
    UIColor* colorFulvous = [UIColor colorWithRed:221.0f/255.0f
                                            green:117.0f/255.0f
                                             blue:0.0f
                                            alpha:1.0f];
    UIColor* colorWhite = [UIColor whiteColor];
    
    UIGraphicsBeginImageContextWithOptions(size, NO, 0.0f);
    
    [colorBlue setFill];
    UIRectFill(CGRectMake(0.0f, 0.0f, size.width, size.height));
    
    [colorFulvous setFill];
    UIBezierPath* path = [UIBezierPath bezierPath];
    [path moveToPoint:CGPointMake(0.0f, size.height / 250.0f * 242.0f)];
    [path addLineToPoint:CGPointMake(size.width, size.height / 250.0f * 53.0f)];
    [path addLineToPoint:CGPointMake(size.width, size.height / 250.0f * 4.0f)];
    [path addLineToPoint:CGPointMake(0.0f, size.height / 250.0f * 238.0f)];
    [path closePath];
    [path fill];

    [colorWhite setFill];
    [path removeAllPoints];
    [path moveToPoint:CGPointMake(0.0f, size.height / 250.0f * 246.0f)];
    [path addLineToPoint:CGPointMake(size.width, size.height / 250.0f * 102.0f)];
    [path addLineToPoint:CGPointMake(size.width, size.height / 250.0f * 53.0f)];
    [path addLineToPoint:CGPointMake(0.0f, size.height / 250.0f * 242.0f)];
    [path closePath];
    [path fill];

    CGFloat radius = size.height / 250.0f * 77.5f;
    CGPoint starCenter = CGPointMake(size.width / 475.0f * 91.27902578, size.height / 250.0f * 91.27902578);
    [[UIBezierPath starAtPoint:starCenter
                        radius:radius
                   innerRadius:radius / 100.0f * 7.7f
                          type:FlagStarTypeFourPointed] fill];
    [[UIBezierPath starAtPoint:starCenter
                        radius:size.height / 250.0f * 55.5f
                   innerRadius:size.height / 250.0f * 24.5f
                          type:FlagStarTypeTwentyFourPointed] fill];
    
    UIImage* image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return image;
}

@end

File: Sources/CountriesFlags/Countries/UIImage+MarshallIslands.swift
//
//  UIImage+MarshallIslands.swift
//  CountriesFlags
//

#if canImport(UIKit)

import UIKit

extension UIImage {
    /**
     [wikipedia]: https://en.wikipedia.org/wiki/Flag_of_the_Marshall_Islands "wikipedia"
     
     Generates the flag of the Marshall Islands.
     For more information see [wikipedia].

     - Parameters:
        - size: image size.
     
     - Returns: Returns the flag of the Marshall Islands.
     */
    internal static func flagMarshallIslands(with size: CGSize) -> UIImage {
        let colorBlue = UIColor(red: 0.0,
                                green: 56.0 / 255.0,
                                blue: 147.0 / 255.0,
                                alpha: 1.0)
        let colorFulvous = UIColor(red: 221.0 / 255.0,
                                   green: 117.0 / 255.0,
                                   blue: 0.0,
                                   alpha: 1.0)
        let colorWhite = UIColor.white

        UIGraphicsBeginImageContextWithOptions(size, true, 0.0)

        colorBlue.setFill()
        UIRectFill(CGRect(origin: .zero, size: size))

        colorFulvous.setFill()
        let path = UIBezierPath()
        path.move(to: CGPoint(x: 0.0, y: size.height / 250.0 * 242.0))
        path.addLine(to: CGPoint(x: size.width, y: size.height / 250.0 * 53.0))
        path.addLine(to: CGPoint(x: size.width, y: size.height / 250.0 * 4.0))
        path.addLine(to: CGPoint(x: 0.0, y: size.height / 250.0 * 238.0))
        path.close()
        path.fill()

        colorWhite.setFill()
        path.removeAllPoints()
        path.move(to: CGPoint(x: 0.0, y: size.height / 250.0 * 246.0))
        path.addLine(to: CGPoint(x: size.width, y: size.height / 250.0 * 102.0))
        path.addLine(to: CGPoint(x: size.width, y: size.height / 250.0 * 53.0))
        path.addLine(to: CGPoint(x: 0.0, y: size.height / 250.0 * 242.0))
        path.close()
        path.fill()

        let radius = size.height / 250.0 * 77.5
        let starCenter = CGPoint(x: size.width / 475.0 * 91.27902578, y: size.height / 250.0 * 91.27902578)
        UIBezierPath.star(at: starCenter,
                          radius: radius,
                          type: .fourPointed(innerRadius: radius / 100.0 * 7.7)).fill()
        UIBezierPath.star(at: starCenter,
                          radius: size.height / 250.0 * 55.5,
                          type: .twentyFourPointed(innerRadius: size.height / 250.0 * 24.5)).fill()

        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()

        return image ?? UIImage()
    }
}

#endif
